clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -783,6 +774,7 @@ void markReconciliationCompleted() {
      * </ol>
      */
     private void resolveMetadataForUnresolvedAssignment() {
+        assignmentReadyToReconcile.clear();
```

Comment:
This call here could make us loose assignments in some cases I expect (we can have assignments ready to reconcile that are not being reconciled yet, and then get more as a result of a metadata update. All need to be kept in the buffer). Assignments ready to reconcile are discovered from 2 places (when we get a HB response, but also when we get Metadata updates), and kept in the buffer if there is a previous reconciliation going on. 

Could you point me to the integration test that you mentioned brought this situation up? I'll take a look to better understand what could be missing.

Reply:
It's the integration test in this PR `testRemoteAssignorRange`.

Reply:
I agree with @lucasbru. I actually had it [here](https://github.com/apache/kafka/pull/14819/files) as well and I thought that it was included in https://github.com/apache/kafka/commit/21edb7078814e5cac41051b170b3ed5ecf4057d5.

Reply:
Oh I see, and agree that we do need the fix (but it was not in https://github.com/apache/kafka/commit/21edb7078814e5cac41051b170b3ed5ecf4057d5, only partly). I was missing the fact that the topics that we may discovered in metadata at any time will be as also included in the new assignments received (if there are still assigned). Basically the same principle that the broker is the source of truth regarding assignment. @lucasbru if you can let the fix in with the test that would be great (or I could just add it in another PR if you prefer). Thanks!

Reply:
Done